tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -270,34 +271,63 @@ std::vector<TensorView*> mlp(
     TensorView* w1,
     TensorView* b1,
     const DeviceMesh& mesh) {
-  DataType dtype = w0->dtype();
-  // Linear #1
-  TensorView* matmul1 = matmul(x, w0);
-  TensorView* b0_bcast = broadcast(b0, {false, true, false});
-  TensorView* linear1 = add(matmul1, b0_bcast);
+  const DataType dtype = w0->dtype();
+  // Linear 0
+  TensorView* matmul0 = matmul(x, w0);
+  TensorView* linear0 = add(matmul0, broadcast(b0, {false, true, false}));
   // GeLU
-  TensorView* gelu = tanh_gelu(linear1);
+  TensorView* gelu = tanh_gelu(linear0);
   gelu = castOp(dtype, gelu);
-  // Linear #2
-  TensorView* local_matmul2 = matmul(gelu, w1);
-  TensorView* matmul2 = sum(local_matmul2, {0}); // Allreduce
-  TensorView* bcast_bias = broadcast(b1, {true, false});
-  TensorView* linear2 = add(matmul2, bcast_bias);
+  // Linear 1
+  TensorView* local_matmul1 = matmul(gelu, w1);
+  TensorView* matmul1 = sum(local_matmul1, {0}); // Allreduce
+  TensorView* linear1 = add(matmul1, broadcast(b1, {true, false}));
   // Dropout
   Val* prob = IrBuilder::create<Val>(1.0 - kDropoutProb);
   Val* scale = IrBuilder::create<Val>(1.0 / (1.0 - kDropoutProb));
-  auto dropout_result = dropout(linear2, prob, scale).output;
+  auto dropout_result = dropout(linear1, prob, scale).output;
 
   // Manual sharding annotations
-  for (auto tv : {x, b1, matmul2, linear2, dropout_result}) {
+  for (auto tv : {x, b1, linear1, dropout_result}) {
     tv->setDeviceMesh(mesh);
   }
-  for (auto tv : {w0, b0, w1, linear1, gelu}) {
+  for (auto tv : {w0, b0, w1, linear0, gelu}) {
     tv->setDeviceMesh(mesh);
     tv->axis(0)->parallelize(ParallelType::DIDx);
   }
 
-  return {linear1, gelu, linear2, dropout_result};
+  return {linear0, gelu, linear1, dropout_result};
+}
+
+std::vector<TensorView*> mha_qkv(
+    TensorView* x,
+    TensorView* w0,
+    TensorView* b0,
+    TensorView* w1,
+    const DeviceMesh& mesh) {
+  DataType dtype = w0->dtype();
+  const auto D = w0->axis(0)->extent()->value().as<int64_t>();
+  // Recompute: linear 0, q, k, and v
```

Comment:
```suggestion
```
I don't find this comment relevant any more after you moved it into this helper function. The word "recompute" only makes sense under the context of backprop.